src/Datadog.Trace/Agent/MovingAverageKeepRateCalculator.cs

```diff
@@ -111,21 +107,23 @@ internal void UpdateBucket()
             var previousDropped = _dropped[index];
             var previousCreated = _created[index];
 
-            var latestDropped = Math.Min(Interlocked.Exchange(ref _latestDrops, 0), uint.MaxValue);
-            var latestCreated = Math.Min(Interlocked.Exchange(ref _latestKeeps, 0) + latestDropped, uint.MaxValue);
+            // Cap at unit.MaxValue events. Very unlikely to reach this value!
+            var latestDropped = (uint)Math.Min(Interlocked.Exchange(ref _latestDrops, 0), uint.MaxValue);
+            var latestCreated = (uint)Math.Min(Interlocked.Exchange(ref _latestKeeps, 0) + latestDropped, uint.MaxValue);
 
-            var totals = _totals;
-            UnpackBits(totals, out var previousTotalDropped, out var previousTotalCreated);
+            var newSumDropped = _sumDrops - previousDropped + latestDropped;
+            var newSumCreated = _sumCreated - previousCreated + latestCreated;
 
-            var newTotalDropped = (uint)Math.Min(((long)previousTotalDropped) - previousDropped + latestDropped, uint.MaxValue);
-            var newTotalCreated = (uint)Math.Min(((long)previousTotalCreated) - previousCreated + latestCreated, uint.MaxValue);
+            var keepRate = newSumCreated == 0
+                               ? 0
+                               : 1 - ((double)newSumDropped / newSumCreated);
 
-            // Pack the totals into a single value we can read and write atomically
-            var packedTotal = PackBits(newTotalDropped, newTotalCreated);
-            Interlocked.Exchange(ref _totals, packedTotal);
+            Interlocked.Exchange(ref _keepRate, keepRate);
```

Comment:
```suggestion
            Volatile.Write(ref _keepRate, keepRate);
```
This will produce the same Interlocked.Exchange on x86, but more optimized code on x64 (unless you care about the results being published immediately, which I doubt)